File: common/src/main/java/net/impleri/playerskills/integration/kubejs/events/SkillsModificationEventJS.java
package net.impleri.playerskills.integration.kubejs.events;

import dev.latvian.mods.kubejs.RegistryObjectBuilderTypes;
import dev.latvian.mods.kubejs.util.ConsoleJS;
import dev.latvian.mods.rhino.util.HideFromJS;
import net.impleri.playerskills.api.Skill;
import net.impleri.playerskills.integration.kubejs.skills.GenericSkillBuilderJS;
import net.impleri.playerskills.registry.RegistryItemNotFound;
import net.impleri.playerskills.server.registry.Skills;
import net.impleri.playerskills.utils.SkillResourceLocation;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.Nullable;

import java.util.Map;
import java.util.function.Consumer;


public class SkillsModificationEventJS extends BaseSkillsRegistryEventJS {
    public SkillsModificationEventJS(Map<String, RegistryObjectBuilderTypes.BuilderType<Skill<?>>> types) {
        super(types);
    }

    public <T> boolean modify(String name, Consumer<GenericSkillBuilderJS<T>> consumer) {


        return modify(name, null, consumer);
    }

    @HideFromJS
    @Nullable
    private <T> Skill<T> getSkill(String skillName) {
        var name = SkillResourceLocation.of(skillName);
        Skill<T> skill = null;
        try {
            skill = Skills.find(name);
            return skill;
        } catch (RegistryItemNotFound e) {
            ConsoleJS.SERVER.error("Unable to find skill " + name);
        }

        return null;
    }

    public <T> boolean modify(String skillName, @Nullable String skillType, Consumer<GenericSkillBuilderJS<T>> consumer) {
        @Nullable Skill<T> skill = getSkill(skillName);
        if (skill == null) {
            return false;
        }

        String type = (skillType != null) ? skillType : skill.getType().toString();
        ResourceLocation name = SkillResourceLocation.of(skillName);

        GenericSkillBuilderJS<T> builder = getBuilder(type, name);
        if (builder == null) {
            return false;
        }

        builder.syncWith(skill);
        consumer.accept(builder);

        Skill<T> newSkill = builder.createObject();
        Skills.upsert(newSkill);
        ConsoleJS.SERVER.info("Updated " + type + " skill " + name);

        return false;
    }

    public <T> boolean remove(String name) {
        @Nullable Skill<T> skill = getSkill(name);
        if (skill == null) {
            return true;
        }

        try {
            Skills.remove(skill);
        } catch (RegistryItemNotFound e) {
            return true;
        }

        ConsoleJS.SERVER.info("Removed skill " + name);

        return true;
    }
}

File: common/src/main/kotlin/net/impleri/playerskills/integrations/kubejs/events/SkillsModificationEventJS.kt
package net.impleri.playerskills.integrations.kubejs.events

import dev.latvian.mods.kubejs.RegistryObjectBuilderTypes
import dev.latvian.mods.kubejs.util.ConsoleJS
import net.impleri.playerskills.api.Skill
import net.impleri.playerskills.integrations.kubejs.skills.GenericSkillBuilderJS
import net.impleri.playerskills.utils.SkillResourceLocation

class SkillsModificationEventJS(types: Map<String, RegistryObjectBuilderTypes.BuilderType<Skill<*>>>) :
  BaseSkillsRegistryEventJS(types) {

  fun <T> modify(
    skillName: String,
    skillType: String?,
    consumer: ((GenericSkillBuilderJS<T>) -> Unit),
  ): Boolean {
    val name = SkillResourceLocation.of(skillName)
    val skill = Skill.find<T>(name) ?: return false

    val type = skillType ?: skill.type.toString()

    val builder = getBuilder<T>(type, name) ?: return false

    builder.syncWith(skill)
    consumer(builder)
    val newSkill = builder.createObject() ?: return false

    val response = Skill.modify(newSkill)
    ConsoleJS.SERVER.info("Updated $type skill $name")

    return response
  }

  fun <T> modify(
    skillName: String,
    consumer: ((GenericSkillBuilderJS<T>) -> Unit),
  ): Boolean {
    return modify(skillName, null, consumer)
  }

  fun <T> remove(name: String): Boolean {
    val skill = Skill.find<T>(name) ?: return true
    val response = Skill.remove(skill)

    ConsoleJS.SERVER.info("Removed skill $name")

    return response
  }
}
